// minecraft-server-util/build/ping.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ping = void 0;
const assert_1 = __importDefault(require("assert"));
const Packet_1 = __importDefault(require("./structure/Packet"));
const TCPSocket_1 = __importDefault(require("./structure/TCPSocket"));
const getVarIntSize_1 = __importDefault(require("./util/getVarIntSize"));
const formatResult_1 = __importDefault(require("./util/formatResult"));
const resolveSRV_1 = __importDefault(require("./util/resolveSRV"));
const getTimeoutPromise_1 = __importDefault(require("./util/getTimeoutPromise"));
const ipAddressRegEx = /^\d{1,3}(\.\d{1,3}){3}$/;
function applyDefaultOptions(options) {
    return Object.assign({
        port: 25565,
        protocolVersion: 47,
        pingTimeout: 1000 * 5,
        enableSRV: true
    }, options);
}
function ping(host, options) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const opts = applyDefaultOptions(options);
        assert_1.default(typeof host === 'string', `Expected 'host' to be a string, got ${typeof host}`);
        assert_1.default(host.length > 0, 'Expected \'host\' to have content, got an empty string');
        assert_1.default(typeof options === 'object' || typeof options === 'undefined', `Expected 'options' to be an object or undefined, got ${typeof options}`);
        assert_1.default(typeof opts === 'object', `Expected 'options' to be an object, got ${typeof opts}`);
        assert_1.default(typeof opts.port === 'number', `Expected 'options.port' to be a number, got ${typeof opts.port}`);
        assert_1.default(opts.port > 0, `Expected 'options.port' to be greater than 0, got ${opts.port}`);
        assert_1.default(opts.port < 65536, `Expected 'options.port' to be less than 65536, got ${opts.port}`);
        assert_1.default(Number.isInteger(opts.port), `Expected 'options.port' to be an integer, got ${opts.port}`);
        assert_1.default(typeof opts.protocolVersion === 'number', `Expected 'options.protocolVersion' to be a number, got ${typeof opts.protocolVersion}`);
        assert_1.default(opts.protocolVersion >= 0, `Expected 'options.protocolVersion' to be greater than or equal to 0, got ${opts.protocolVersion}`);
        assert_1.default(Number.isInteger(opts.protocolVersion), `Expected 'options.protocolVersion' to be an integer, got ${opts.protocolVersion}`);
        assert_1.default(typeof opts.pingTimeout === 'number', `Expected 'options.pingTimeout' to be a number, got ${typeof opts.pingTimeout}`);
        assert_1.default(opts.pingTimeout > 0, `Expected 'options.pingTimeout' to be greater than 0, got ${opts.pingTimeout}`);
        assert_1.default(typeof opts.enableSRV === 'boolean', `Expected 'options.enableSRV' to be a boolean, got ${typeof opts.enableSRV}`);
        let srvRecord = null;
        if (opts.enableSRV && !ipAddressRegEx.test(host)) {
            srvRecord = yield resolveSRV_1.default(host);
        }
        const socket = yield TCPSocket_1.default.connect((_a = srvRecord === null || srvRecord === void 0 ? void 0 : srvRecord.host) !== null && _a !== void 0 ? _a : host, opts.port, opts.pingTimeout);
        {
            const handshakePacket = new Packet_1.default();
            handshakePacket.writeVarInt(0x00);
            handshakePacket.writeVarInt(opts.protocolVersion);
            handshakePacket.writeString(host, true);
            handshakePacket.writeShort(opts.port);
            handshakePacket.writeVarInt(1);
            socket.writePacket(handshakePacket, true);
            const requestPacket = new Packet_1.default();
            requestPacket.writeVarInt(0x00);
            socket.writePacket(requestPacket, true);
        }
        let data = null;
        for (let i = 0; i < 3; i++) {
            const packetLength = yield socket.readVarInt();
            const packetType = yield socket.readVarInt();
            if (packetType !== 0) {
                const readSize = packetLength - getVarIntSize_1.default(packetType);
                if (readSize > 0) {
                    yield socket.readBytes(readSize);
                }
                continue;
            }
            data = yield socket.readString();
            break;
        }
        yield socket.destroy();
        if (data === null) {
            throw new Error('Failed to recieve correct packet within 3 attempts');
        }
        let result;
        try {
            result = JSON.parse(data);
        }
        catch (e) {
            throw new Error('Response from server is not valid JSON');
        }
        return formatResult_1.default(host, opts.port, srvRecord, result);
    });
}
function pingWithTimeout(host, options) {
    var _a;
    return Promise.race([
        ping(host, options),
        getTimeoutPromise_1.default((_a = options === null || options === void 0 ? void 0 : options.pingTimeout) !== null && _a !== void 0 ? _a : 1000 * 15, 'Failed to ping server within time')
    ]);
}
exports.ping = pingWithTimeout;
