// minecraft-server-util/build/pingFE01FA.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pingFE01FA = void 0;
const assert_1 = __importDefault(require("assert"));
const Packet_1 = __importDefault(require("./structure/Packet"));
const TCPSocket_1 = __importDefault(require("./structure/TCPSocket"));
const formatResultFE01FA_1 = __importDefault(require("./util/formatResultFE01FA"));
const resolveSRV_1 = __importDefault(require("./util/resolveSRV"));
const decodeUTF16BE_1 = __importDefault(require("./util/decodeUTF16BE"));
const getTimeoutPromise_1 = __importDefault(require("./util/getTimeoutPromise"));
const ipAddressRegEx = /^\d{1,3}(\.\d{1,3}){3}$/;
function applyDefaultOptions(options) {
    return Object.assign({
        port: 25565,
        protocolVersion: 47,
        pingTimeout: 1000 * 5,
        enableSRV: true
    }, options);
}
function pingFE01FA(host, options) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const opts = applyDefaultOptions(options);
        assert_1.default(typeof host === 'string', `Expected 'host' to be a string, got ${typeof host}`);
        assert_1.default(host.length > 0, 'Expected \'host\' to have content, got an empty string');
        assert_1.default(typeof options === 'object' || typeof options === 'undefined', `Expected 'options' to be an object or undefined, got ${typeof options}`);
        assert_1.default(typeof opts === 'object', `Expected 'options' to be an object, got ${typeof opts}`);
        assert_1.default(typeof opts.port === 'number', `Expected 'options.port' to be a number, got ${typeof opts.port}`);
        assert_1.default(opts.port > 0, `Expected 'options.port' to be greater than 0, got ${opts.port}`);
        assert_1.default(opts.port < 65536, `Expected 'options.port' to be less than 65536, got ${opts.port}`);
        assert_1.default(Number.isInteger(opts.port), `Expected 'options.port' to be an integer, got ${opts.port}`);
        assert_1.default(typeof opts.protocolVersion === 'number', `Expected 'options.protocolVersion' to be a number, got ${typeof opts.protocolVersion}`);
        assert_1.default(opts.protocolVersion >= 0, `Expected 'options.protocolVersion' to be greater than or equal to 0, got ${opts.protocolVersion}`);
        assert_1.default(Number.isInteger(opts.protocolVersion), `Expected 'options.protocolVersion' to be an integer, got ${opts.protocolVersion}`);
        assert_1.default(typeof opts.pingTimeout === 'number', `Expected 'options.pingTimeout' to be a number, got ${typeof opts.pingTimeout}`);
        assert_1.default(opts.pingTimeout > 0, `Expected 'options.pingTimeout' to be greater than 0, got ${opts.pingTimeout}`);
        assert_1.default(typeof opts.enableSRV === 'boolean', `Expected 'options.enableSRV' to be a boolean, got ${typeof opts.enableSRV}`);
        let srvRecord = null;
        if (opts.enableSRV && !ipAddressRegEx.test(host)) {
            srvRecord = yield resolveSRV_1.default(host);
        }
        const socket = yield TCPSocket_1.default.connect((_a = srvRecord === null || srvRecord === void 0 ? void 0 : srvRecord.host) !== null && _a !== void 0 ? _a : host, opts.port, opts.pingTimeout);
        {
            const packet = new Packet_1.default();
            packet.writeByte(0xFE, 0x01, 0xFA, 0x00, 0x0B, 0x00, 0x4D, 0x00, 0x43, 0x00, 0x7C, 0x00, 0x50, 0x00, 0x69, 0x00, 0x6E, 0x00, 0x67, 0x00, 0x48, 0x00, 0x6F, 0x00, 0x73, 0x00, 0x74);
            packet.writeShort(7 + host.length);
            packet.writeByte(opts.protocolVersion);
            packet.writeShort(host.length);
            packet.writeString(host, false);
            packet.writeInt(opts.port);
            socket.writePacket(packet, false);
        }
        let protocolVersion = 0;
        let serverVersion = '';
        let motd = '';
        let playerCount = 0;
        let maxPlayers = 0;
        {
            const packetType = yield socket.readByte();
            if (packetType !== 0xFF) {
                throw new Error('Packet returned from server was unexpected type');
            }
            const length = yield socket.readShort();
            const data = decodeUTF16BE_1.default(String.fromCodePoint(...(yield socket.readBytes(length * 2)).slice(6)));
            const [protocolVersionStr, serverVersionStr, motdStr, playerCountStr, maxPlayersStr] = data.split('\0');
            protocolVersion = parseInt(protocolVersionStr);
            serverVersion = serverVersionStr;
            motd = motdStr;
            playerCount = parseInt(playerCountStr);
            maxPlayers = parseInt(maxPlayersStr);
            if (isNaN(protocolVersion)) {
                throw new Error('Server returned an invalid protocol version: ' + protocolVersionStr);
            }
            if (isNaN(playerCount)) {
                throw new Error('Server returned an invalid player count: ' + playerCountStr);
            }
            if (isNaN(maxPlayers)) {
                throw new Error('Server returned an invalid max player count: ' + maxPlayersStr);
            }
        }
        yield socket.destroy();
        return formatResultFE01FA_1.default(host, opts.port, srvRecord, protocolVersion, serverVersion, motd, playerCount, maxPlayers);
    });
}
function pingWithTimeout(host, options) {
    var _a;
    return Promise.race([
        pingFE01FA(host, options),
        getTimeoutPromise_1.default((_a = options === null || options === void 0 ? void 0 : options.pingTimeout) !== null && _a !== void 0 ? _a : 1000 * 15, 'Failed to ping server within time')
    ]);
}
exports.pingFE01FA = pingWithTimeout;
