// minecraft-server-util/build/query.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.query = void 0;
const assert_1 = __importDefault(require("assert"));
const Packet_1 = __importDefault(require("./structure/Packet"));
const resolveSRV_1 = __importDefault(require("./util/resolveSRV"));
const getTimeoutPromise_1 = __importDefault(require("./util/getTimeoutPromise"));
const UDPSocket_1 = __importDefault(require("./structure/UDPSocket"));
const parseDescription_1 = __importDefault(require("./util/parseDescription"));
const ipAddressRegEx = /^\d{1,3}(\.\d{1,3}){3}$/;
let sessionCounter = 0;
function applyDefaultOptions(options) {
    return Object.assign({
        port: 25565,
        queryTimeout: 1000 * 5,
        enableSRV: true,
        sessionID: ++sessionCounter
    }, options);
}
function query(host, options) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const opts = applyDefaultOptions(options);
        assert_1.default(typeof host === 'string', `Expected 'host' to be a string, got ${typeof host}`);
        assert_1.default(host.length > 0, 'Expected \'host\' to have content, got an empty string');
        assert_1.default(typeof options === 'object' || typeof options === 'undefined', `Expected 'options' to be an object or undefined, got ${typeof options}`);
        assert_1.default(typeof opts === 'object', `Expected 'options' to be an object, got ${typeof opts}`);
        assert_1.default(typeof opts.port === 'number', `Expected 'options.port' to be a number, got ${typeof opts.port}`);
        assert_1.default(opts.port > 0, `Expected 'options.port' to be greater than 0, got ${opts.port}`);
        assert_1.default(opts.port < 65536, `Expected 'options.port' to be less than 65536, got ${opts.port}`);
        assert_1.default(Number.isInteger(opts.port), `Expected 'options.port' to be an integer, got ${opts.port}`);
        assert_1.default(typeof opts.queryTimeout === 'number', `Expected 'options.queryTimeout' to be a number, got ${typeof opts.queryTimeout}`);
        assert_1.default(opts.queryTimeout > 0, `Expected 'options.queryTimeout' to be greater than 0, got ${opts.queryTimeout}`);
        assert_1.default(typeof opts.sessionID === 'number', `Expected 'options.sessionID' to be a number, got ${typeof opts.sessionID}`);
        assert_1.default(opts.sessionID > 0, `Expected 'options.sessionID' to be greater than 0, got ${opts.sessionID}`);
        assert_1.default(opts.sessionID < 0xFFFFFFFF, `Expected 'options.sessionID' to be less than ${0xFFFFFFFF}, got ${opts.sessionID}`);
        assert_1.default(Number.isInteger(opts.sessionID), `Expected 'options.sessionID' to be an integer, got ${opts.sessionID}`);
        assert_1.default(typeof opts.enableSRV === 'boolean', `Expected 'options.enableSRV' to be a boolean, got ${typeof opts.enableSRV}`);
        let challengeToken;
        let srvRecord = null;
        if (opts.enableSRV && !ipAddressRegEx.test(host)) {
            srvRecord = yield resolveSRV_1.default(host);
        }
        const socket = new UDPSocket_1.default((_a = srvRecord === null || srvRecord === void 0 ? void 0 : srvRecord.host) !== null && _a !== void 0 ? _a : host, opts.port);
        {
            const requestPacket = new Packet_1.default();
            requestPacket.writeByte(0xFE, 0xFD, 0x09);
            requestPacket.writeInt(opts.sessionID);
            yield socket.writePacket(requestPacket);
        }
        {
            const responsePacket = yield socket.readPacket();
            const type = responsePacket.readByte();
            const sessionID = responsePacket.readInt();
            challengeToken = parseInt(responsePacket.readStringNT());
            if (type !== 0x09) {
                throw new Error('Server sent an invalid payload type');
            }
            if (sessionID !== opts.sessionID) {
                throw new Error('Session ID in response did not match client session ID');
            }
            if (isNaN(challengeToken)) {
                throw new Error('Server sent an invalid challenge token');
            }
        }
        {
            const requestPacket = new Packet_1.default();
            requestPacket.writeByte(0xFE, 0xFD, 0x00);
            requestPacket.writeInt(opts.sessionID);
            requestPacket.writeInt(challengeToken);
            yield socket.writePacket(requestPacket);
        }
        let motd, gameType, levelName, onlinePlayers, maxPlayers;
        {
            const responsePacket = yield socket.readPacket();
            const type = responsePacket.readByte();
            const sessionID = responsePacket.readInt();
            motd = responsePacket.readStringNT();
            gameType = responsePacket.readStringNT();
            levelName = responsePacket.readStringNT();
            const onlinePlayersStr = responsePacket.readStringNT();
            const maxPlayersStr = responsePacket.readStringNT();
            if (type !== 0x00) {
                throw new Error('Server sent an invalid payload type');
            }
            if (sessionID !== opts.sessionID) {
                throw new Error('Session ID in response did not match client session ID');
            }
            onlinePlayers = parseInt(onlinePlayersStr);
            if (isNaN(onlinePlayers)) {
                throw new Error('Server sent an invalid player count');
            }
            maxPlayers = parseInt(maxPlayersStr);
            if (isNaN(maxPlayers)) {
                throw new Error('Server sent an invalid max player count');
            }
        }
        yield socket.destroy();
        return {
            host,
            port: opts.port,
            srvRecord,
            gameType,
            levelName,
            onlinePlayers,
            maxPlayers,
            description: parseDescription_1.default(motd)
        };
    });
}
function queryWithTimeout(host, options) {
    var _a;
    return Promise.race([
        query(host, options),
        getTimeoutPromise_1.default((_a = options === null || options === void 0 ? void 0 : options.queryTimeout) !== null && _a !== void 0 ? _a : 1000 * 15, 'Failed to query server within time')
    ]);
}
exports.query = queryWithTimeout;
