// minecraft-server-util/build/structure/TCPSocket.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const net_1 = __importDefault(require("net"));
const Packet_1 = __importDefault(require("./Packet"));
class TCPSocket {
    constructor(socket) {
        this.isConnected = false;
        this.buffer = [];
        this.socket = socket;
        socket.on('data', (data) => {
            this.buffer.push(...Array.from(data));
        });
    }
    static connect(host, port, timeout) {
        assert_1.default(host.length > 0, 'Expected host.length > 0, got ' + host.length);
        assert_1.default(Number.isInteger(port), 'Expected integer, got ' + port);
        assert_1.default(port > 0, 'Expected port > 0, got ' + port);
        assert_1.default(port < 65536, 'Expected port < 65536, got ' + port);
        assert_1.default(timeout > 0, 'Expected timeout > 0, got ' + timeout);
        const socket = net_1.default.createConnection({ host, port, timeout });
        return new Promise((resolve, reject) => {
            const connectHandler = () => {
                resolve(new TCPSocket(socket));
                socket.removeListener('connect', connectHandler);
            };
            const closeHandler = () => {
                reject();
                socket.removeListener('close', closeHandler);
            };
            const endHandler = () => {
                reject();
                socket.removeListener('end', endHandler);
            };
            const errorHandler = (error) => {
                reject(error);
                socket.removeListener('error', errorHandler);
            };
            const timeoutHandler = () => {
                reject();
                socket.removeListener('timeout', timeoutHandler);
            };
            socket.on('connect', connectHandler);
            socket.on('close', closeHandler);
            socket.on('end', endHandler);
            socket.on('error', errorHandler);
            socket.on('timeout', timeoutHandler);
        });
    }
    readByte() {
        if (this.buffer.length > 0) {
            return Promise.resolve(this.buffer.shift() || 0);
        }
        return new Promise((resolve) => {
            let read = false;
            const dataHandler = () => {
                if (read) {
                    return;
                }
                process.nextTick(() => {
                    if (this.buffer.length > 0) {
                        read = true;
                        this.socket.removeListener('data', dataHandler);
                        return resolve(this.buffer.shift());
                    }
                });
            };
            this.socket.on('data', dataHandler);
        });
    }
    readBytes(length) {
        if (this.buffer.length >= length) {
            const value = this.buffer.slice(0, length);
            this.buffer.splice(0, length);
            return Promise.resolve(value);
        }
        return new Promise((resolve) => {
            let read = false;
            const dataHandler = () => {
                if (read) {
                    return;
                }
                process.nextTick(() => {
                    if (this.buffer.length >= length) {
                        read = true;
                        this.socket.removeListener('data', dataHandler);
                        const value = this.buffer.slice(0, length);
                        this.buffer.splice(0, length);
                        return resolve(value);
                    }
                });
            };
            this.socket.on('data', dataHandler);
        });
    }
    readVarInt() {
        return __awaiter(this, void 0, void 0, function* () {
            let numRead = 0;
            let result = 0;
            let read, value;
            do {
                if (numRead > 4) {
                    throw new Error('VarInt exceeds data bounds');
                }
                read = yield this.readByte();
                value = (read & 0b01111111);
                result |= (value << (7 * numRead));
                numRead++;
                if (numRead > 5) {
                    throw new Error('VarInt is too big');
                }
            } while ((read & 0b10000000) != 0);
            return result;
        });
    }
    readString() {
        return __awaiter(this, void 0, void 0, function* () {
            const length = yield this.readVarInt();
            const data = yield this.readBytes(length);
            let value = '';
            for (let i = 0; i < data.length; i++) {
                value += String.fromCharCode(data[i]);
            }
            return value;
        });
    }
    readShort() {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.readBytes(2);
            return (data[0] << 8) | data[1];
        });
    }
    readInt() {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.readBytes(4);
            return (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3];
        });
    }
    readIntLE() {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.readBytes(4);
            return data[0] | (data[1] << 8) | (data[2] << 16) | (data[3] << 24);
        });
    }
    readLong() {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.readBytes(8);
            return (data[0] << 56) | (data[1] << 48) | (data[2] << 40) | (data[3] << 32) | (data[4] << 24) | (data[5] << 16) | (data[6] << 8) | data[7];
        });
    }
    writeByte(value) {
        return new Promise((resolve, reject) => {
            this.socket.write(Buffer.from([value]), (error) => {
                if (error) {
                    return reject(error);
                }
                resolve();
            });
        });
    }
    writeBytes(buffer) {
        return new Promise((resolve, reject) => {
            this.socket.write(buffer, (error) => {
                if (error) {
                    return reject(error);
                }
                resolve();
            });
        });
    }
    writePacket(packet, prefixLength) {
        if (prefixLength) {
            const finalPacket = new Packet_1.default();
            finalPacket.writeVarInt(packet.data.length);
            finalPacket.writeByte(...packet.data);
            return this.writeBytes(Buffer.from(finalPacket.data));
        }
        return this.writeBytes(Buffer.from(packet.data));
    }
    destroy() {
        return new Promise((resolve) => {
            this.socket.removeAllListeners();
            this.socket.end(() => {
                resolve();
                this.socket.destroy();
            });
        });
    }
}
exports.default = TCPSocket;
